tools/block-generator/generator/generator_ledger.go

```diff
@@ -167,19 +170,30 @@ func (g *generator) startEvaluator(hdr bookkeeping.BlockHeader, paysetHint int)
 		})
 }
 
-func (g *generator) evaluateBlock(hdr bookkeeping.BlockHeader, txGroups [][]txn.SignedTxnWithAD, paysetHint int) (*ledgercore.ValidatedBlock, uint64 /* txnCount */, error) {
+func (g *generator) evaluateBlock(hdr bookkeeping.BlockHeader, txGroups [][]txn.SignedTxnWithAD, paysetHint int) (*ledgercore.ValidatedBlock, uint64 /* txnCount */, time.Duration /* commit wait time */, error) {
+	commitWaitTime := time.Duration(0)
+	waitDelay := 10 * time.Millisecond
 	eval, err := g.startEvaluator(hdr, paysetHint)
 	if err != nil {
-		return nil, 0, fmt.Errorf("could not start evaluator: %w", err)
+		return nil, 0, 0, fmt.Errorf("could not start evaluator: %w", err)
 	}
 	for i, txGroup := range txGroups {
-		err := eval.TransactionGroup(txGroup)
-		if err != nil {
-			return nil, 0, fmt.Errorf("could not evaluate transaction group %d: %w", i, err)
+		for {
+			err := eval.TransactionGroup(txGroup)
+			if err != nil {
+				if strings.Contains(err.Error(), "database table is locked") {
+					time.Sleep(waitDelay)
+					commitWaitTime += waitDelay
+					// sometimes the database is locked, so we retry
+					continue
+				}
+				return nil, 0, 0, fmt.Errorf("could not evaluate transaction group %d: %w", i, err)
+			}
+			break
```

Comment:
This is the "bug fix". The error always seems to happen on the first transaction group. Simply retrying until it works seems to allow the tests to complete.

Reply:
Or maybe more informative would be a `retriedBlocks` counter. So a contribution of 1 for blocks which had to retry, and 0 for those without any retries.

Reply:
I was mainly looking for magnitude with this. In my first test the importer was waiting for ~5 seconds over the course of a 1h test, so I don't think we need to worry about the wait time too much.